polkadot/runtime/common/src/paras_registrar/mod.rs

```diff
@@ -692,6 +912,125 @@ impl<T: Config> OnNewHead for Pallet<T> {
 	}
 }
 
+impl<T: Config> PreCodeUpgrade for Pallet<T> {
+	/// Ensures that all upgrade-related costs are covered for the specific parachain. Upon success,
+	/// it updates the deposit-related state for the parachain.
+	///
+	/// There are three cases where we do not charge any upgrade costs from the initiator
+	/// of the upgrade:
+	///
+	/// 1. `requirements` is explicitly set to `UpgradeRequirements::SkipRequirements`.
+	///
+	/// 2. System parachains do not pay for upgrade fees.
+	///
+	/// 3. All lease-holding parachains are permitted to do upgrades for free. This is introduced to
+	///    avoid causing a breaking change to the system once para upgrade fees are required.
+	fn pre_code_upgrade(
+		para: ParaId,
+		new_code: ValidationCode,
+		requirements: UpgradeRequirements,
+	) -> Result<Weight, Weight> {
+		let Some(head) = paras::Pallet::<T>::para_head(para) else {
+			return Err(T::DbWeight::get().reads(1))
+		};
+
+		let Some(mut info) = Paras::<T>::get(para) else { return Err(T::DbWeight::get().reads(2)) };
+
+		let new_deposit = Self::required_para_deposit(head.0.len(), new_code.0.len());
+		let current_deposit = info.deposit;
+
+		let lease_holding = Self::is_parachain(para);
+		let free_upgrade = requirements == UpgradeRequirements::SkipRequirements ||
+			para.is_system() ||
+			lease_holding;
+
+		if !free_upgrade {
+			let Some(billing_account) = info.billing_account.clone() else {
+				Self::deposit_event(Event::<T>::CodeUpgradeScheduleFailed(
+					CodeUpgradeScheduleError::BillingAccountNotSet,
+				));
+				// An overestimate of the used weight, but it's better to be safe than sorry.
+				return Err(<T as Config>::WeightInfo::pre_code_upgrade())
+			};
+
+			if let Err(_) = <T as Config>::Currency::withdraw(
+				&billing_account,
+				T::UpgradeFee::get(),
+				WithdrawReasons::FEE,
+				ExistenceRequirement::KeepAlive,
+			) {
+				Self::deposit_event(Event::<T>::CodeUpgradeScheduleFailed(
+					CodeUpgradeScheduleError::FailedToPayUpgradeFee,
+				));
+				// An overestimate of the used weight, but it's better to be safe than sorry.
+				return Err(<T as Config>::WeightInfo::pre_code_upgrade())
+			}
+
+			let additional_deposit = new_deposit.saturating_sub(current_deposit);
+			if let Err(_) = <T as Config>::Currency::reserve(&billing_account, additional_deposit) {
+				Self::deposit_event(Event::<T>::CodeUpgradeScheduleFailed(
+					CodeUpgradeScheduleError::FailedToReserveDeposit,
+				));
+				// An overestimate of the used weight, but it's better to be safe than sorry.
+				return Err(<T as Config>::WeightInfo::pre_code_upgrade())
```

Comment:
Do we on purpose not refund the `withdrawn` fee above? If not, we could run both in a storage transaction and revert both if one failed.

Reply:
Made this transactional, should make more sense now.